keps/sig-storage/20191031-storage-capacity-constraints-for-pod-scheduling.md

```diff
@@ -0,0 +1,935 @@
+`--enable-capacity=topology-fallback`, external-provisioner
+determines storage pools based on the topology information provided by
+the driver on each node and gets capacity for the resulting pools from
+the driver's controller service that it is connected to.
+
+This relies on a feature of kubelet and one additional requirement for CSI
+driver deployments:
+- Every topology segment that a CSI node service reports in response
+ to GetNodeInfo (for example, `example.com/zone`: `Z1` +
+ `example.com/rack`: `R3`) is copied into node labels by kubelet.
+- All keys must have the same prefix (`example.com` in that example)
+  and external-provisioner is given that prefix with the
+  `--topology-prefix=example.com` parameter.
+
+With that parameter, external-provisioner can reconstruct topology segments as follows:
+- For each node, find the labels with that prefix and combine them
+  into a `csi.Topology` instance.
+- Remove duplicates.
+
+When `topology-ephemeral` is active, for each storage class that uses
+the driver and for each of the topology segments it calls
+`GetCapacity`, with parameters from the storage class and the topology
+segment as `GetCapacityRequest.accessible_topology`.
+
+When `topology-ephemeral` is active, it also calls
+`GetCapacity` for the same segment without parameters. In other words, capacity
+information for ephemeral inline volumes is gathered through the CSI
+controller service and is assumed to be specific to the same topology
+segments as normal volumes.
+
+As before, `topology-fallback` covers the case where storage class
+parameters do not affect capacity.
+
+The result is then stored in a different `CSIStoragePool` object for
+each identified pool, with a `NodeTopology` that selects the right
+nodes via their labels.
+
+Drivers that don't support topology can use
+`--enable-capacity=global-ephemeral`,
+`--enable-capacity=global-persistent` and/or
+`--enable-capacity=global-fallback`. Those are similar to their
+`topology-*` counterparts but without identifying pools and thus
+without `GetCapacityRequest.accessible_topology` The resulting
+`CSIStoragePool` objects then have neither `Nodes` nor `NodeSelector` set and
+thus apply to all nodes in the cluster.
+
+#### CSIStoragePool lifecycle
+
+external-provisioner needs permission to create, update and delete
+`CSIStoragePool` objects. Before creating a new object, it must check
+whether one already exists with the relevant attributes (driver name +
+nodes) and then update that one instead.
+
+To ensure that `CSIStoragePool` objects get removed when the driver
+deployment gets removed before it gets a chance to clean up, each
+`CSIStoragePool` object needs an [owner
+reference](https://godoc.org/k8s.io/apimachinery/pkg/apis/meta/v1#OwnerReference)
+that points to the pod which runs the external-provisioner that
+created it. Then those objects will get garbage collected.
+
+It is an intentional side effect that this happens also when the pod
```

Comment:
Instead of the pod, what about the controller that created it, ie DaemonSet, StatefulSet, Deployment? Then only when the driver gets uninstalled, the objects will get cleaned up.

Reply:
This was intentional. For example, take node-local storage where each external-provisioner instance in a DaemonSet is responsible for one CSIStoragePool object. If one such instance is removed permanently, for example after a node failure, then we want to also have that CSIStoragePool removed.

The same if a StatefulSet instance suddently dies: we don't want to have the information that it was responsible for hanging around any longer than necessary, because then the scheduler will act based on stale information.

Reply:
In the non-local case, this will cause downtime of provisioning during potentially normal operations like upgrading the driver, where the provisioner already has leader-election enabled to avoid these kinds of downtime. I think in the general case, it would be better to have an ownerRef to the controller rather than the specific pod. The node case could even have an ownerRef to the Node object instead of the Pod object. Pods may not get cleaned up immediately in certain node failure scenarios (another can of worms for another kep)

Reply:
See 86cf87dc7a2e58fcfec9034180d14ef6bab29fcb.